Tools/scripts/freeze_modules.py

```diff
@@ -0,0 +1,499 @@
+    return lines[:start_pos + 1] + replacements + lines[end_pos:]
+
+
+def regen_frozen(specs, dest=MODULES_DIR):
+    if isinstance(dest, str):
+        frozen, frozenids = resolve_frozen_files(specs, destdir)
+    else:
+        frozenids, frozen = dest
+
+    headerlines = []
+    parentdir = os.path.dirname(FROZEN_FILE)
+    for frozenid in frozenids:
+        # Adding a comment to separate sections here doesn't add much,
+        # so we don't.
+        _, frozenfile = frozen[frozenid]
+        header = os.path.relpath(frozenfile, parentdir)
+        headerlines.append(f'#include "{header}"')
+
+    deflines = []
+    indent = '    '
+    lastsection = None
+    for spec in specs:
+        frozenid, _, modname, ispkg, section = spec
+        if section != lastsection:
+            if lastsection is not None:
+                deflines.append('')
+            deflines.append(f'/* {section} */')
+        lastsection = section
+
+        # This matches what we do in Programs/_freeze_module.c:
+        name = frozenid.replace('.', '_')
+        symbol = '_Py_M__' + name
+        pkg = '-' if ispkg else ''
+        line = ('{"%s", %s, %s(int)sizeof(%s)},'
+                % (modname, symbol, pkg, symbol))
+        # TODO: Consider not folding lines
+        if len(line) < 80:
+            deflines.append(line)
+        else:
+            line1, _, line2 = line.rpartition(' ')
+            deflines.append(line1)
+            deflines.append(indent + line2)
+
+    if not deflines[0]:
+        del deflines[0]
+    for i, line in enumerate(deflines):
+        if line:
+            deflines[i] = indent + line
+
+    with updating_file_with_tmpfile(FROZEN_FILE) as (infile, outfile):
+        lines = infile.readlines()
+        # TODO: Use more obvious markers, e.g.
+        # $START GENERATED FOOBAR$ / $END GENERATED FOOBAR$
+        lines = replace_block(
+            lines,
+            "/* Includes for frozen modules: */",
+            "/* End includes */",
+            headerlines,
+            FROZEN_FILE,
+        )
+        lines = replace_block(
+            lines,
+            "static const struct _frozen _PyImport_FrozenModules[] =",
+            "/* sentinel */",
+            deflines,
+            FROZEN_FILE,
+        )
+        outfile.writelines(lines)
+
+
+def regen_makefile(frozenids, frozen):
+    frozenfiles = []
+    rules = ['']
+    for frozenid in frozenids:
+        pyfile, frozenfile = frozen[frozenid]
+        header = os.path.relpath(frozenfile, ROOT_DIR)
+        # Adding a comment to separate sections here doesn't add much,
+        # so we don't.
+        relfile = header.replace('\\', '/')
+        frozenfiles.append(f'\t\t$(srcdir)/{relfile} \\')
+
+        _pyfile = os.path.relpath(pyfile, ROOT_DIR)
+        tmpfile = f'{header}.new'
+        # Note that we freeze the module to the target .h file
+        # instead of going through an intermediate file like we used to.
```

Comment:
If this is considered a problem (I think the scenario was where a crash in the generator script would leave behind an empty output file) we can add something like you did in update_file.py to _freeze_module.c. But personally I think that generator is rock-solid, so we probably don't need to bother.

Reply:
Yeah, I don't think we need to bother.